lib/plugins/dartdoc/plugin.dart

```diff
@@ -0,0 +1,346 @@
+
+  for (var i = 0; i <= n; i++) {
+    v0[i] = i;
+  }
+
+  for (var i = 0; i <= m - 1; i++) {
+    var previousV1 = v1[0] = i + 1;
+    var previousV0 = v0[0];
+
+    for (var j = 0; j <= n - 1; j++) {
+      var substitutionCost = previousV0 + 1;
+      if (s[i] == t[j]) {
+        substitutionCost--;
+      }
+
+      previousV0 = v0[j + 1];
+
+      final deletionCost = previousV0;
+      final insertionCost = previousV1;
+
+      var minInsertionDeletion =
+          deletionCost < insertionCost ? deletionCost : insertionCost;
+      minInsertionDeletion++;
+
+      final overallMin = substitutionCost < minInsertionDeletion
+          ? substitutionCost
+          : minInsertionDeletion;
+
+      previousV1 = overallMin;
+      v1[j + 1] = previousV1;
+    }
+
+    (v0, v1) = (v1, v0);
+  }
+
+  return v0[n];
+}
+
+/// Listens for search requests in messages sent in a Discord channel, processes
+/// them, and sends the results back to the channel.
+///
+/// Searches can be embedded in any text message and their syntax is as follows:
+/// - `![Name]` or `![package/Name]`: Return the documentation for `Name` in
+///   Flutter's or `package`'s API documentation.
+/// - `?[Name]` or `?[package/Name]`: Search for `Name` in Flutter's or
+///   `package`'s API documentation.
+/// - `\$[name]`: Return the pub.dev page for the package `name`.
+/// - `&[name]`: Search pub.dev for `name`.
+class DartdocSearch extends NyxxPlugin<NyxxGateway> {
+  /// The time after which cached documentation entries are considered to have
+  /// expired and require refreshing.
+  static const expireDuration = Duration(hours: 1);
+
+  @override
+  String get name => 'DartdocSearch';
+
+  final _documentationCache = <String, (DateTime, List<DartdocEntry>)>{};
+  late final Timer _flutterCacheTimer;
+
+  /// Create a new [DartdocSearch] and begin populating the documentation cache
+  /// with Flutter's documentation.
+  DartdocSearch() {
+    // Fetching the flutter cache index takes quite a long time (as it is a very
+    // large package) and it is probably the most commonly used package, so we
+    // eagerly update the cache before it expires.
+    //
+    // Other packages are fetched and updated as needed.
+    _flutterCacheTimer = Timer.periodic(
+      expireDuration * 0.9,
+      (_) => unawaited(getEntries('flutter')),
+    );
+
+    // Kick off an initial load immediately
+    unawaited(getEntries('flutter'));
+  }
+
+  /// Return a record of `(url, entries)` for the given `package`.
+  ///
+  /// `url` is the URL relative to which the [DartdocEntry.href] should be
+  /// resolved.
+  /// `entries` is a list of [DartdocEntry] in the package's documentation.
+  Future<(String, List<DartdocEntry>)> getEntries(String package) async {
+    final cached = _documentationCache[package];
+    final now = DateTime.timestamp();
+    final cacheExpiry = now.add(-expireDuration);
+
+    final urlBase = package == 'flutter' || package == 'dart'
```

Comment:
Do you think we'll use the Flutter docs commonly enough for this to be worthwhile over just the Dart docs? It's quite big.

Reply:
The issue is that on the off chance someone _does_ request a flutter API element, it takes a noticeable amount of time (~10 seconds on my machine and internet, albeit both of which are slow) to fetch & parse the API elements. And once someone does request it, it's in the cache until the bot is restarted. The only advantage of not prefetching it is saving on the ~31MB of network traffic it generates slightly over once per hour.

So I don't see much point in not prefetching the flutter docs. What could however be interesting for us is to link to https://api.dart.dev instead of https://api.flutter.dev for SDK API elements, to avoid confusing users and default to only searching the SDK API, so we only point to https://api.flutter.dev if explicitely requested.

What do you think?

Reply:
I see, thanks for the explanation! I'm fine with still caching and surfacing the Flutter docs, but it would be nice to link to api.dart.dev for Dart SDK elements if that's possible.

Reply:
This is only an issue with Flutter because its API is so huge. Most packages on pub.dev are fetched fast enough it isn't noticeable.

Reply:
> It would definitely be nice to link to api.dart.dev for Dart SDK elements if that's possible, but not a big deal if it's a pain.

I was about to say it would be simple, but I don't see a way to always get the latest `index.json` from `api.dart.dev`. The current versions is at https://api.dart.dev/stable/3.2.1/index.json, but that has the version 3.2.1 encoded into it.
For some reason, while a URL like `https://api.dart.dev/stable/dart-math/dart-math-library.html` redirects to `https://api.dart.dev/stable/3.2.1/dart-math/dart-math-library.html`, `https://api.dart.dev/stable/index.json` redirects you to `https://api.dart.dev/stable/3.2.1/index.html`.

Reply:
Yeah, I had a similar problem when looking into flame docs some time ago. I was following a tutorial and at the same time searching for the API and classes of the package. Always getting different results. I suggested them to create a latest path and always redirect there and avoid search bots from indexing the other versions. Maybe we could suggest something similar to Dart and Flutter if that ever becomes an issue